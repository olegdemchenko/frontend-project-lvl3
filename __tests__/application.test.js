import '@testing-library/jest-dom';
import { screen } from '@testing-library/dom';
import fs from 'fs';
import path from 'path';
import prettier from 'prettier';
import userEvent from '@testing-library/user-event';
import nock from 'nock';
import axios from 'axios';
import timer from 'timer-promise';
import run from '../src/app/application';
import getUrlWithCORSFree from '../src/common';

axios.defaults.adapter = require('axios/lib/adapters/http');

nock.disableNetConnect();

const options = {
  parser: 'html',
  htmlWhitespaceSensitivity: 'ignore',
  tabWidth: 4,
};

const getFormattedHTML = () => prettier.format(document.body.innerHTML, options);
const pathToFixture = (fixtureName) => path.join('__tests__', '__fixtures__', fixtureName);
const readFile = (filename) => fs.readFileSync(pathToFixture(filename)).toString();

const links = {
  notValid: 'hello world',
  notExisted: 'http://somenotexistedsite.com',
  rss: 'http://lorem-rss.herokuapp.com',
};

let elements;
beforeEach(() => {
  const initHtml = readFile('index.html');
  document.body.innerHTML = initHtml;
  elements = {
    submit: screen.getByRole('button'),
    emailInput: screen.getByRole('textbox', { name: /url/ }),
  };
  return run.then((app) => app());
});

test('validation error', () => (
  userEvent.type(elements.emailInput, links.notValid)
    .then(() => userEvent.click(elements.submit))
    .then(() => timer.start('start', 15))
    .then(() => expect(getFormattedHTML()).toMatchSnapshot())
));

test('response error', () => {
  const scope = nock(getUrlWithCORSFree(links.notExisted)).get('/rss').reply(404);
  return userEvent.type(elements.emailInput, `${links.notExisted}/rss`)
    .then(() => userEvent.click(elements.submit))
    .then(() => timer.start('start', 200))
    .then(() => {
      expect(getFormattedHTML()).toMatchSnapshot();
      scope.done();
    });
});

test('success', () => {
  const response = readFile('lorem.html');
  const scope = nock(getUrlWithCORSFree(links.rss)).get('/feed').reply(200, response);
  return userEvent.type(elements.emailInput, `${links.rss}/feed`)
    .then(() => userEvent.click(elements.submit))
    .then(() => timer.start('start', 200))
    .then(() => {
      expect(getFormattedHTML()).toMatchSnapshot();
      scope.done();
    });
});

test('update', () => {
  const responseBeforeUpdate = readFile('loremBeforeUpdate.html');
  const responseAfterUpdate = readFile('loremAfterUpdate.html');
  const scope = nock(getUrlWithCORSFree(links.rss))
    .get('/feed?unit=second')
    .reply(200, responseBeforeUpdate)
    .get('/feed?unit=second')
    .reply(200, responseAfterUpdate);
  return userEvent.type(elements.emailInput, `${links.rss}/feed?unit=second`)
    .then(() => userEvent.click(elements.submit))
    .then(() => timer.start('start', 5500))
    .then(() => {
      expect(getFormattedHTML()).toMatchSnapshot();
      scope.done();
    });
});
