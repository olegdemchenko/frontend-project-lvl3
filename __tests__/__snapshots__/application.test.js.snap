// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`success 1`] = `
"<main class=\\"flex-grow-1\\">
    <section class=\\"jumbotron jumbotron-fluid bg-dark\\">
        <div class=\\"container-xl\\">
            <div class=\\"row\\">
                <div class=\\"col-md-10 col-lg-8 mx-auto text-white\\">
                    <h1 class=\\"display-3\\">RSS Reader</h1>
                    <p class=\\"lead\\">
                        Start reading RSS today! Is is easy, it is nicely.
                    </p>
                    <form action=\\"\\" class=\\"rss-form\\">
                        <div class=\\"form-row\\">
                            <div class=\\"col\\">
                                <input
                                    class=\\"form-control form-control-lg w-100\\"
                                    type=\\"text\\"
                                    autofocus=\\"autofocus\\"
                                    required=\\"required\\"
                                    name=\\"url\\"
                                    aria-label=\\"url\\"
                                    placeholder=\\"RSS link\\"
                                />
                            </div>
                            <div class=\\"col-auto\\">
                                <button
                                    name=\\"button\\"
                                    aria-label=\\"add\\"
                                    class=\\"btn btn-lg btn-primary px-sm-5\\"
                                >
                                    Add
                                </button>
                            </div>
                        </div>
                    </form>
                    <p class=\\"text-muted my-1\\">
                        Example: https://ru.hexlet.io/lessons.rss
                    </p>
                    <div class=\\"feedback text-success\\" data-testid=\\"feedback\\">
                        rss has been loaded
                    </div>
                </div>
            </div>
        </div>
    </section>
    <div class=\\"container-xl\\">
        <div class=\\"row\\">
            <div class=\\"col-md-10 col-lg-8 mx-auto feeds\\">
                <ul>
                    <h2>Feeds list</h2>
                    <h3>
                        <a href=\\"http://example.com/\\">
                            Lorem ipsum feed for an interval of 1 minutes with
                            10 item(s)
                        </a>
                    </h3>
                </ul>
                <ul>
                    <h2>
                        Lorem ipsum feed for an interval of 1 minutes with 10
                        item(s)
                    </h2>
                    <li>
                        <a href=\\"http://example.com/test/1597855200\\">
                            Lorem ipsum 2020-08-19T16:40:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597855140\\">
                            Lorem ipsum 2020-08-19T16:39:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597855080\\">
                            Lorem ipsum 2020-08-19T16:38:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597855020\\">
                            Lorem ipsum 2020-08-19T16:37:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597854960\\">
                            Lorem ipsum 2020-08-19T16:36:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597854900\\">
                            Lorem ipsum 2020-08-19T16:35:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597854840\\">
                            Lorem ipsum 2020-08-19T16:34:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597854780\\">
                            Lorem ipsum 2020-08-19T16:33:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597854720\\">
                            Lorem ipsum 2020-08-19T16:32:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597854660\\">
                            Lorem ipsum 2020-08-19T16:31:00Z
                        </a>
                    </li>
                </ul>
            </div>
        </div>
    </div>
</main>
<footer class=\\"footer border-top py-3 mt-5\\">
    <div class=\\"container-xl\\">
        <div class=\\"text-center\\">Created By Oleg Demchenko</div>
    </div>
</footer>
"
`;

exports[`update 1`] = `
"<main class=\\"flex-grow-1\\">
    <section class=\\"jumbotron jumbotron-fluid bg-dark\\">
        <div class=\\"container-xl\\">
            <div class=\\"row\\">
                <div class=\\"col-md-10 col-lg-8 mx-auto text-white\\">
                    <h1 class=\\"display-3\\">RSS Reader</h1>
                    <p class=\\"lead\\">
                        Start reading RSS today! Is is easy, it is nicely.
                    </p>
                    <form action=\\"\\" class=\\"rss-form\\">
                        <div class=\\"form-row\\">
                            <div class=\\"col\\">
                                <input
                                    class=\\"form-control form-control-lg w-100\\"
                                    type=\\"text\\"
                                    autofocus=\\"autofocus\\"
                                    required=\\"required\\"
                                    name=\\"url\\"
                                    aria-label=\\"url\\"
                                    placeholder=\\"RSS link\\"
                                />
                            </div>
                            <div class=\\"col-auto\\">
                                <button
                                    name=\\"button\\"
                                    aria-label=\\"add\\"
                                    class=\\"btn btn-lg btn-primary px-sm-5\\"
                                >
                                    Add
                                </button>
                            </div>
                        </div>
                    </form>
                    <p class=\\"text-muted my-1\\">
                        Example: https://ru.hexlet.io/lessons.rss
                    </p>
                    <div class=\\"feedback text-success\\" data-testid=\\"feedback\\">
                        rss has been loaded
                    </div>
                </div>
            </div>
        </div>
    </section>
    <div class=\\"container-xl\\">
        <div class=\\"row\\">
            <div class=\\"col-md-10 col-lg-8 mx-auto feeds\\">
                <ul>
                    <h2>Feeds list</h2>
                    <h3>
                        <a href=\\"http://example.com/\\">
                            Lorem ipsum feed for an interval of 1 minutes with
                            10 item(s)
                        </a>
                    </h3>
                </ul>
                <ul>
                    <h2>
                        Lorem ipsum feed for an interval of 1 minutes with 10
                        item(s)
                    </h2>
                    <li>
                        <a href=\\"http://example.com/test/1597926540\\">
                            Lorem ipsum 2020-08-20T12:29:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597926480\\">
                            Lorem ipsum 2020-08-20T12:28:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597926420\\">
                            Lorem ipsum 2020-08-20T12:27:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597926360\\">
                            Lorem ipsum 2020-08-20T12:26:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597926300\\">
                            Lorem ipsum 2020-08-20T12:25:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597926240\\">
                            Lorem ipsum 2020-08-20T12:24:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597926180\\">
                            Lorem ipsum 2020-08-20T12:23:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597926120\\">
                            Lorem ipsum 2020-08-20T12:22:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597926060\\">
                            Lorem ipsum 2020-08-20T12:21:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597926000\\">
                            Lorem ipsum 2020-08-20T12:20:00Z
                        </a>
                    </li>
                    <li>
                        <a href=\\"http://example.com/test/1597925940\\">
                            Lorem ipsum 2020-08-20T12:19:00Z
                        </a>
                    </li>
                </ul>
            </div>
        </div>
    </div>
</main>
<footer class=\\"footer border-top py-3 mt-5\\">
    <div class=\\"container-xl\\">
        <div class=\\"text-center\\">Created By Oleg Demchenko</div>
    </div>
</footer>
"
`;
